packages/table/src/common/internal/selectionUtils.ts

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+import { IRegion, RegionCardinality, Regions } from "../../regions";
+import { IFocusedCellCoordinates } from "../cell";
+import { Direction } from "../direction";
+import * as DirectionUtils from "./directionUtils";
+import * as FocusedCellUtils from "./focusedCellUtils";
+
+/**
+ * Resizes the provided region by 1 row/column in the specified direction,
+ * returning a new region instance. The region may either expand *or* contract
+ * depending on the presence and location of the focused cell.
+ *
+ * If no focused cell is provided, the region will always be *expanded* in the
+ * specified direction.
+ *
+ * If a focused cell *is* provided, the behavior will change depending on where
+ * the focused cell is within the region:
+ *
+ *   1. If along a top/bottom boundary while resizing UP/DOWN, the resize will
+ *      expand from or shrink to the focused cell (same if along a left/right
+ *      boundary while moving LEFT/RIGHT).
+ *   2. If *not* along a top/bottom boundary while resizing UP/DOWN (or if *not*
+ *      along a left/right boundary while moving LEFT/RIGHT), the region will
+ *      simply expand in the specified direction.
+ *
+ * Other notes:
+ * - A CELLS region can be resized vertically or horizontally.
+ * - A FULL_ROWS region can be resized only vertically.
+ * - A FULL_COLUMNS region can be resized only horizontally.
+ * - A FULL_TABLE region cannot be resized.
+ *
+ * This function does not clamp the indices of the returned region; that is the
+ * responsibility of the caller.
+ */
+export function resizeSelectedRegion(region: IRegion, direction: Direction, focusedCell?: IFocusedCellCoordinates) {
```

Comment:
How are these terms: `resize` (used on shift + arrow keys) vs `expand` (used on shift + click). Note that they behave differently when the focused cell is floating within the selection (due to <kbd>Tab</kbd>ing or <kbd>Enter</kbd>ing)! Would love to figure out some elegant way to marry all this logic and do away with the need for two names in the first place.

Shift + arrow keys:
![2017-10-12 00 52 08](https://user-images.githubusercontent.com/443450/31485142-a49d778a-aee7-11e7-8dc3-d63ae381edc8.gif)

Shift + click:
![2017-10-12 00 52 23](https://user-images.githubusercontent.com/443450/31485145-a8b67362-aee7-11e7-8637-eff06b8762c9.gif)